module/zfs/dmu_object.c

```diff
@@ -409,6 +409,8 @@ dmu_object_next(objset_t *os, uint64_t *objectp, boolean_t hole, uint64_t txg)
 		 * hand off to dnode_next_offset() for further scanning.
 		 */
 		while (i <= last_obj) {
+			if (i == 0)
+				return (SET_ERROR(ESRCH));
```

Comment:
The fix makes sense, but why not check for this earlier instead of in the while loop?  Maybe something like:

```c
        if (*objectp + 1 == 0) {
                return (SET_ERROR(ESRCH));
        } else if (*objectp == 0) {
                start_obj = 1;
        } else if (ds && dsl_dataset_feature_is_active(ds,
```

Reply:
My concern was that if for some reason we had a dataset where some of the quota objects (which take up slots -3, -2,-1) were missing but not others, or if we got here in the hole case where it would iterate over each one individually looking for a gap, then we could end up getting to the i == 0 case during the loop even if we didn't start there.

Reply:
I see, thanks for the explanation.